File: 3_cxx_Accelerate/lib/accelerate/accelerate.swift
//===----------------------------------------------------------------------===//
//
// This source file is part of the Swift open source project
//
// Copyright (c) 2023 Apple Inc. and the Swift project authors.
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
//
//===----------------------------------------------------------------------===//

import Accelerate

public struct Indices {
   var entries : [Int32] = []
 
   public init()
   {}

public func get(_ i : Int) -> Int32
   {
     return entries[i]
   }
 
   public mutating func push_back(_ x: Int32)
   {
     entries.append(x)
   }
 }

public struct Values {
   var entries : [Double] = []
 
   public init()
   {}

   public init(_ size : Int , _ ignore: Int)
   {
     self.entries =  [Double] (repeating: 0.0, count: size)
   }
   
   public mutating func set(_ i : Int, _ d : Double)
   {
    entries[i] = d;
   }
   
   public func get(_ i : Int) -> Double
   {
     return entries[i]
   }

   public mutating func push_back(_ x: Double)
   {
     entries.append(x)
   }
   public func size() -> Int
   {
     return entries.count;
   }
 }



public func solve(_ dimension : Int32, _ rowIndices : inout Indices, _ columnIndices: inout Indices, _ aValues: inout Values, _ bValues:  inout Values, _ solution : inout Values ) {
  isolve(dimension, rowIndices.entries, columnIndices.entries, aValues.entries, bValues.entries, &solution.entries)
}





public struct Matrix {
  var dimension : Int32
  var A : SparseMatrix_Double
  var factorization : SparseOpaqueFactorization_Double

  public init()
  {
    dimension = 0
    A = SparseMatrix_Double()
    factorization = SparseOpaqueFactorization_Double()
  }

  public mutating func initialize( _ dimension : Int32, _ rowIndices: inout Indices, _ columnIndices: inout Indices, _ aValues: inout Values)
  {
    self.dimension = dimension
    A = SparseConvertFromCoordinate(dimension, dimension,
                                     aValues.entries.count, 1,
                                     SparseAttributes_t(),
                                     rowIndices.entries, columnIndices.entries,
                                     aValues.entries)
    
    factorization = SparseFactor(SparseFactorizationQR, A)
  }


  // deinit
  // {}


  public func solve( _ bValues:  inout Values, _ solution : inout Values)
  {
    solution.entries = bValues.entries.map { $0 }


    /// Solve the system.
    solution.entries.withUnsafeMutableBufferPointer { sPtr in
      let xb = DenseVector_Double(count: dimension,
                                  data: sPtr.baseAddress!)
    
       SparseSolve(factorization, xb)
   }
 }
}



func isolve(_ dimension : Int32, _ rowIndices: [Int32], _ columnIndices: [Int32], _ aValues: [Double], _ bValues:  [Double], _ solution : inout [Double] ) {

  let A = SparseConvertFromCoordinate(dimension, dimension,
                                      aValues.count, 1,
                                      SparseAttributes_t(),
                                      rowIndices, columnIndices,
                                      aValues)
 /// Factorize _A_.
  let factorization = SparseFactor(SparseFactorizationQR, A)

  defer {
    SparseCleanup(A)
    SparseCleanup(factorization)
  }

  solution = bValues.map { $0 }

  /// Solve the system.
  solution.withUnsafeMutableBufferPointer { sPtr in
    let xb = DenseVector_Double(count: dimension,
                                data: sPtr.baseAddress!)
    
    SparseSolve(factorization, xb)
  }
}

File: 3_cxx_Accelerate/src/accelerate.cpp
//===----------------------------------------------------------------------===//
//
// This source file is part of the Swift open source project
//
// Copyright (c) 2023 Apple Inc. and the Swift project authors.
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
//
//===----------------------------------------------------------------------===//

#include "accelerate/accelerate-swift.h"
#include <iostream>

int main(int argc, char ** argv) {


  SwiftAccelerate::Indices rows = SwiftAccelerate::Indices::init();
  SwiftAccelerate::Indices cols = SwiftAccelerate::Indices::init();
  SwiftAccelerate::Values values = SwiftAccelerate::Values::init();
  SwiftAccelerate::Values B = SwiftAccelerate::Values::init();
  SwiftAccelerate::Values X = SwiftAccelerate::Values::init();

  rows.push_back(0);
  cols.push_back(2);
  values.push_back(10.0);

  rows.push_back(1);
  cols.push_back(0);
  values.push_back(20.0);

  rows.push_back(1);
  cols.push_back(2);
  values.push_back(5.0);

  rows.push_back(2);
  cols.push_back(1);
  values.push_back(50.0);

  B.push_back(30.0);
  B.push_back(35.0);
  B.push_back(100.0);

  X.push_back(0.0);
  X.push_back(0.0);
  X.push_back(0.0);

  SwiftAccelerate::solve(3, rows, cols, values, B, X);
  
  std::cout << X.get(0) << std::endl;
  std::cout << X.get(1) << std::endl;
  std::cout << X.get(2) << std::endl;


  std::cout << "Now use Matrix" << std::endl;

  SwiftAccelerate::Matrix M = SwiftAccelerate::Matrix::init(3, rows, cols, values);
  M.solve(B, X);
  std::cout << X.get(0) << std::endl;
  std::cout << X.get(1) << std::endl;
  std::cout << X.get(2) << std::endl;
  
  return 0;
}
